librad/src/net/protocol/io/graft.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{collections::BTreeSet, iter, net::SocketAddr, str::FromStr, time::Duration};
+
+use crate::{
+    git::{
+        fetch::{Fetcher as _, RemoteHeads},
+        refs::{self, Refs, Remotes},
+        replication,
+        storage::{self, fetcher},
+        Urn,
+    },
+    PeerId,
+};
+
+pub mod error {
+    use super::*;
+    use thiserror::Error;
+
+    #[derive(Debug, Error)]
+    #[non_exhaustive]
+    pub enum Rere {
+        #[error("precomputed signed refs for {0} not found")]
+        MissingSignedRefs(Urn),
+
+        #[error(transparent)]
+        Replicate(#[from] replication::Error),
+
+        #[error(transparent)]
+        Refs(#[from] refs::stored::Error),
+
+        #[error("unable to obtain fetcher")]
+        Fetcher(#[from] fetcher::error::Retrying<git2::Error>),
+
+        #[error(transparent)]
+        Storage(#[from] storage::Error),
+
+        #[error(transparent)]
+        Pool(#[from] storage::PoolError),
+    }
+}
+
+pub mod config {
+    use super::*;
+
+    pub struct Rere {
+        pub replication: replication::Config,
+        pub fetch_slot_wait_timeout: Duration,
+    }
+}
+
+/// Initiate [`replication::replicate`] if the `remote_peer` appears to serve
+/// "interesting" refs.
+///
+/// Remote refs are deemed interesting if there is an intersection between the
+/// set of [`PeerId`]s acc. to the local [`Remotes`] for the given [`Urn`], and
+/// the advertised remote tracking branches of the remote peer (plus the remote
+/// peer itself).
+///
+/// This function is invoked when the `remote_peer` initiates a fetch (hence
+/// "rere" -- replicate replicate), as per [`super::recv::git`].
+///
+/// A fetch initiated by this function will **not** generate a
+/// [`crate::git::p2p::header::Header::nonce`], so as to not trigger reres
+/// recursively. Using this function thus requires to inspect the git header for
```

Comment:
Can you expand on this "inspect the git header" part? Is it something the caller does? Do we have an example of it?

Reply:
I linked to [`super::recv::git`], is that not enough?

Reply:
Right, I missed that :+1:

Reply:
The question I wanted you to ask was: why is it that the nonce is set by default, and unset when `rere`ing, vs. unset by default and set by `rere`.

The answer is that I'm not sure either.

Reply:
Ok, now you've hooked me. I think I need a bit more prose on the `nonce`. I _thought_ I understood its purpose but now I've unconvinced myself.

Reply:
Maybe you can say what you thought?

What's documented here is that, since we trigger the `rere` by just a regular fetch and not some special-fetch, we need to be able to distinguish which one it is regardless, otherwise we'd just `rere` until eternity. It is possible that a fetch doesn't trigger a `rere` at all (because not interesting), or that the `rere` comes in after the original fetch has completed (very likely, actually).

So, we need to correlate `re` and `rere` pairs over some time window, right?

Things to keep in mind:

* The feature may get rolled out to a network which doesn't know `rere`, but is otherwise compatible
* If one side doesn't adhere to the protocol, can it do harm to the other side?
* Are there reasons to opt-out of the feature, and if so, which end should be able to decide?
* What other uses of a unique "request identifier" can you think of?

Reply:
> Maybe you can say what you thought?

I understood that it was marking when to mutually replicate but didn't quite piece it together, so I took another look and I'll try to explain what I think is happening.

The `Header` is the advertisement of what git action to do. We're embedding the `nonce` into the `Header` to advertise to the remote that we're in a mutual fetch mode. So `p1` will initiate a `replicate` from `p2` and will use a `nonce`. `p2` will see this `nonce` and decide if any of the remotes that `p1` has are interesting to it [`p2`]. If `p2` sees something interesting then it will also initiate a `replicate` but _will not_ send a `nonce` otherwise we would just keep going around in circles.

Admittedly, your explanation in the thread helped too.

> * The feature may get rolled out to a network which doesn't know rere, but is otherwise compatible
> * If one side doesn't adhere to the protocol, can it do harm to the other side?

These make sense, and from what I understand the `nonce` being optional satisfies these requirements. If the `nonce` is missing the other mutual fetch never happens but the `remote_peer` still at least gives the peer replicating the data it requested.

> * Are there reasons to opt-out of the feature, and if so, which end should be able to decide?

Again, a `nonce` could be ignored by intercepting the header and setting it `None`, right?

> * What other uses of a unique "request identifier" can you think of?

Heh, good question... None that spring to mind but that doesn't necessarily mean there won't be any. So maybe we shouldn't paint ourselves into a corner with the `nonce` thinking it'll be the only `nonce`. What I mean is that `n=` is now representing mutual fetches. But if we extend our cases and we need another `nonce` how do we choose the next parameter names and differentiate them?

One other question, [GitService::run](https://github.com/radicle-dev/radicle-link/pull/569/files#diff-310c08d56ff78cfbcec33b5d1c52979878d05c7f337eeb53b130ce606d655c26R101) ignores the `nonce`. Why is that? :thinking:

Reply:
> Again, a nonce could be ignored by intercepting the header and setting it None, right?

Uhm, you need to BYO fetcher now, which you can tell to not set the nonce. Since `replicate` is part of the public API, library users can say "fetch me this, but tell the other end to not dial back".

> Again, a nonce could be ignored by intercepting the header and setting it None, right?

We can stick more parameters at the end of the header line (`\0` separated). But since this is just a random value, I'm not sure what you have in mind by "another nonce".

A common pattern around nonces is that you expect the other end to "echo" it back to you, which I thought could come in handy.

> One other question, GitService::run ignores the nonce. Why is that?

Maybe you missed that I split the invocation of the "git server" in two steps now? You first call `service` which parses the header and returns you a type from which you can inspect it. Then, you actually run the thing -- which doesn't have any business with the `n` parameter, but needs to know the `service` and `repo` from the header.

Reply:
> not sure what you have in mind by "another nonce".

So if we had something like:
```rust
GitUrl {
   /// mutual fetch nonce
   nonce: Option<u32>,
   /// that new feature that we introduced that also required a nonce
   new_feature_nonce: Option<u32>,
}
```

> Maybe you missed that I split the invocation of the "git server" in two steps now? You first call service which parses the header and returns you a type from which you can inspect it. Then, you actually run the thing -- which doesn't have any business with the n parameter, but needs to know the service and repo from the header.

Ah yes, I see that now :+1: Something that could aid with connecting the concepts is some documentation on the `nonce` fields because their construction and their use are in different parts of the code base.

This has all been super helpful for understanding :blush:

Reply:
> /// that new feature that we introduced that also required a nonce

Yes sure, but why would we introduce another _random_ value? I mean, we could figure out that we want a monotonically increasing value instead, in which case it might be better to encode `n` now such that it doesn’t look like a number, so we could distinguish it. Or we could figure out that we want `proofofwork=256bits` instead, in which case we’d probably deprecate `n`, because that’s obviously superior entropy.

Reply:
> "fetch me this, but tell the other end to not dial back"

Yeah, so, Mr. Duck, you convinced me that we actually need to keep track of both sent and seen nonces (which I was trying to avoid). Otherwise, if you present a fresh nonce every time, you can keep the other side busy. If it is only `rere` which sends the nonce, it can keep the other side busy by just not sending a nonce. Instead, `rere` needs to echo the nonce it saw. Unnonced requests should get throttled. Thanks

Reply:
Or maybe I should not think about such things later in the day.

Reply:
> Yes sure, but why would we introduce another random value?

I can't predict the future, unfortunately :smile: This was just a postulation of another case happening to pop up.

> we actually need to keep track of both sent and seen nonces

So this would say, "hey I've seen a nonce from this peer recently, and I answered so you can piss off I'm not gonna `rere`, come back tomorrow"?